service/history/queueProcessor.go

```diff
@@ -0,0 +1,361 @@
+
+	p.shutdownWG.Add(1)
+	p.NotifyNewTask()
+	go p.processorPump()
+}
+
+func (p *queueProcessorBase) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.isStopped, 0, 1) {
+		return
+	}
+
+	logging.LogQueueProcesorShuttingDownEvent(p.logger)
+	defer logging.LogQueueProcesorShutdownEvent(p.logger)
+
+	if atomic.LoadInt32(&p.isStarted) == 1 {
+		close(p.shutdownCh)
+	}
+
+	if success := common.AwaitWaitGroup(&p.shutdownWG, time.Minute); !success {
+		logging.LogQueueProcesorShutdownTimedoutEvent(p.logger)
+	}
+}
+
+func (p *queueProcessorBase) NotifyNewTask() {
+	var event struct{}
+	select {
+	case p.notifyCh <- event:
+	default: // channel already has an event, don't block
+	}
+}
+
+func (p *queueProcessorBase) processorPump() {
+	defer p.shutdownWG.Done()
+	tasksCh := make(chan queueTaskInfo, p.options.BatchSize)
+
+	var workerWG sync.WaitGroup
+	for i := 0; i < p.options.WorkerCount; i++ {
+		workerWG.Add(1)
+		go p.taskWorker(tasksCh, &workerWG)
+	}
+
+	pollTimer := time.NewTimer(p.options.MaxPollInterval)
+	updateAckTimer := time.NewTimer(p.options.UpdateAckInterval)
+
+processorPumpLoop:
+	for {
+		select {
+		case <-p.shutdownCh:
+			break processorPumpLoop
+		case <-p.notifyCh:
+			p.processBatch(tasksCh)
+		case <-pollTimer.C:
+			p.processBatch(tasksCh)
+			pollTimer = time.NewTimer(p.options.MaxPollInterval)
+		case <-updateAckTimer.C:
+			p.ackMgr.updateAckLevel()
+			updateAckTimer = time.NewTimer(p.options.UpdateAckInterval)
+		}
+	}
+
+	p.logger.Info("Queue processor pump shutting down.")
+	// This is the only pump which writes to tasksCh, so it is safe to close channel here
+	close(tasksCh)
+	if success := common.AwaitWaitGroup(&workerWG, 10*time.Second); !success {
+		p.logger.Warn("Queue processor timed out on worker shutdown.")
+	}
+	updateAckTimer.Stop()
+	pollTimer.Stop()
+}
+
+func (p *queueProcessorBase) processBatch(tasksCh chan<- queueTaskInfo) {
+
+	if !p.rateLimiter.Consume(1, p.options.MaxPollInterval) {
+		p.NotifyNewTask() // re-enqueue the event
+		return
+	}
+
+	tasks, err := p.ackMgr.readTransferTasks()
+
+	if err != nil {
+		p.logger.Warnf("Processor unable to retrieve tasks: %v", err)
+		p.NotifyNewTask() // re-enqueue the event
+		return
+	}
+
+	if len(tasks) == 0 {
+		return
+	}
+
+	for _, tsk := range tasks {
+		tasksCh <- tsk
+	}
+
+	if len(tasks) == p.options.BatchSize {
+		// There might be more task
+		// We return now to yield, but enqueue an event to poll later
+		p.NotifyNewTask()
+	}
+
+	return
+}
+
+func (p *queueProcessorBase) taskWorker(tasksCh <-chan queueTaskInfo, workerWG *sync.WaitGroup) {
+	defer workerWG.Done()
+
+	for {
+		select {
+		case task, ok := <-tasksCh:
+			if !ok {
+				return
+			}
+
+			p.processWithRetry(task)
+		}
+	}
+}
+
+func (p *queueProcessorBase) processWithRetry(task queueTaskInfo) {
+	p.logger.Debugf("Processing task: %v, type: %v", task.GetTaskID(), task.GetTaskType())
+ProcessRetryLoop:
+	for retryCount := 1; retryCount <= 100; retryCount++ {
```

Comment:
retryCount should be configurable, ie. a const / var or input parameter

Reply:
Good point.  Let me make this change.